### PR TITLE
EIP-156: Reclaiming of Ether in common classes of stuck accounts

### LINKED ISSUE #156: Reclaiming of ether in common classes of stuck accounts
### Specification (v1)

The below is only usable between blocks `FORK_BLKNUM` and `FORK_BLKNUM + EFFECTIVE_DURATION` (both TBA).

If the `v` value of a signature is (strictly) greater than 1024, then calculate the sender as follows:
1. Let `s` be the sender computed according to the normal algorithm, using 27 as the `v` value if the provided `v` is odd, and 28 if the provided `v` is even.
2. Let the actual sender be the address that a contract would be created at if its sender is `s` and the contract creation nonce is `floor((v - 1025) / 2)`.

Transactions with `v` values strictly greater than 1024 are only valid if the sender account is nonexistent or its code is empty.

If the `v` value of a signature is 1023 or 1024, then calculate the sender as follows:
1. Let `P` be the **public key** computed according to the normal algorithm, in the 64-byte packed form that is normally hashed to determine the sender address, using 27 as the `v` value if the provided `v` is odd, and 28 if the provided `v` is even.
2. Let the actual sender be the last 20 bytes of the sha3 of the lowercase non-prefixed hex encoded form of `P` instead of the binary raw `P` itself.

### Specification (v2)

Create a solidity contract with the following functions:

* `declareEmptyContract(index)`: computes `rlp.encode([msg.sender, index])`; if there is a contract at this address with ether and no code, then the contract saves a record that this contract has been checked, and sends an equal amount of "future ether" (an ERC20 token) at that account.
* `declareLowercaseHexAddress(pubkey)`: checks `sha3(pubkey) % 2**160 == msg.sender`; then checks that `sha3(pubkey.encode('hex')) % 2**160` has ether; if it does, then the contract saves a record that this contract has been checked, and sends an equal amount of "future ether" (an ERC20 token) at that account
* `withdraw()`: deletes the `msg.sender`'s future ether, and sends it an equivalent amount of ether.

The hard fork would increase the future ether contract's balance by an amount equal to the total quantity of extant future ether.

### Specification (v3)

Follow v2, but distribute the future ether on a Casper testnet. Then, later have a single step that converts both Casper ether and ethereum 1.0 ether into ether as part of the Serenity hardfork.

### Rationale

This allows for users with ether or other assets in common classes of "stuck" accounts to withdraw their assets. The first case covers contracts that are accidentally created with no code, as well as some losses due to replay attacks where a contract was created on ETC, funds sent on ETH but the contract not created on ETH; the second case covers losses due to an old ethereum javascript library that incorrectly computed ethereum addresses. Note that in all cases, the "rightful owner" of the assets is obvious and mathematically provable, and no user is being deprived of any assets, and this proposal provides no explicit favor to any single account, user or application.

It is understood that there may be a risk that this proposal will be viewed controversially as it is in some sense a "rescue" rather than a "technical improvement", even though it is arguably much less intrusive than previous such proposals for the reasons outlined above; the proposal is created in order to allow community discussion and debate and does not signify full endorsement.

### PR DESCRIPTION
**Pull request by 5chdn**
Closes #156 

I'm stepping up as _champion_ for EIP-156. Kindly paging @vbuterin. 

Included proposals:

- This specifies 156A (aka "v1" - Contracts without code).
- This specifies 156B (aka "v2" - Incorrectly computed addresses).

Not included proposals:

- This does not specify "[v3](https://github.com/ethereum/EIPs/issues/156#issue-183007625)" as I do not see any or improvements advantages over "v2" (but feel free to prove me wrong).
- This does not specify 156C (aka "[parity multi-sig library](https://github.com/ethereum/EIPs/issues/156#issuecomment-342748195)") due to a potential conflict of interest. (See https://github.com/5chdn/EIPs/pull/1 for preview and discussion.)
- This does not specify 156D (aka "[cosmos crowdsale-bug](https://github.com/ethereum/EIPs/issues/156#issuecomment-344700108)") due to lack of a working specification. (See https://github.com/5chdn/EIPs/pull/2 for preview and discussion.)

Notes:

- I suggested to include 156A,B in the Constantinople hard-fork and renamed the variables accordingly.
- I suggested a one-year claim period for 156A, renamed the variable subsequently, and recommended two million blocks.
- ~~This is early work-in-progress, I still need to catch up with the full discussion in #156, but I feel this draft needs to be out early.~~ Done.
- I'm happy to append this proposal by 156C and 156D as discussed in the linked comments if this is generally desired and technically well-specified. Kindly paging @holiman and @ebuchman.

Let's try to split the discussion to keep an overview: general concerns and political implications go into #156; technical aspects go into #803 (here). Thanks.

---

Disclosure: I have no assets in any locked accounts. Work on this EIP is a personal effort and not in the scope of employment by Parity Technologies.

### CODE REVIEW

**Comment by holiman** [comment]
The cosmos crowdsale bug has a working implementation; see https://github.com/ethereum/EIPs/issues/156#issuecomment-345000373

**Comment by 5chdn** [comment]
@holiman right, thanks. I started adding this https://github.com/5chdn/EIPs/pull/2/files#diff-7beb254da2e418cbc42ca6ba03b1c3e6R57

Need to expand on the _calculates the corresponding cosmos address_ part, yet.

**Comment by holiman** [comment]
This is the part which calculates the cos address: 
```
    function testCosRecovery() constant returns (bytes20){
        bytes memory a = hex"02D7485091B95C158EEEC8ED2E248F7164BE5DF0956583DB3BB99F326390081CF8";
        var s = sha256(a);
        var b = ripemd160(s);
        // returns 0x4fb8fab66ef33955f10d59b9b86183148858dbce
        return b;
    }
```

And `a` represents the compressed pubkey.

**Comment by gcolvin** [comment]
I've put a link for discussion here as well. https://gitter.im/ethereum/ether-recovery?at=5a3a88c119147ac3231ae130

**Comment by MicahZoltu** [comment]
This EIP adds a bunch of really unnecessary complexity.  If recovering lost ETH is desired, then we can simply include a list of accounts that need to have one-time state transitions applied to them as part of a hard fork.  The Ethereum client implementation complexity of one-time state transitions for a certain set of accounts is _way_ lower than adding a new set of temporary EVM rules.

If desired, the logic proposed in this EIP could be used to decide which accounts get added to the list and which don't, and that logic can be executed off-chain and validated by anyone prior to the fork.

**Comment by gcolvin** [comment]
Makes sense to me @MicahZoltu.  What transactions besides the ones that fit the rules here might be on the list?

File: EIPS/eip-156.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+The hard fork would increase the future Ether contract's balance after `CONSTANTINOPLE_FORK_BLKNUM` by an amount equal to the total quantity of extant _future Ether_.
+
+## Rationale
+It is understood that there may be a risk that this proposal will be viewed controversially as it is in some sense a _rescue_ rather than a _technical improvement_, even though it is arguably much less intrusive than previous such proposals for the reasons outlined above.
+
+## Backwards Compatibility
+This proposal is not forward compatible and introduces backward incompatibilities in the set of valid transactions (156A) and a non-standard state-transition (156B).
+
+## Test Cases
+-156A: None.
```
Comment (pirapira) [inline EIPS/eip-156.md]: A space is missing between `-` and `156A`.

File: EIPS/eip-156.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
+## Preamble
+
+    EIP: 156
+    Title: Reclaiming of Ether in common classes of stuck accounts
+    Author: Vitalik Buterin, Afri Schoedon
+    Type: Standard Track
+    Category: Core
+    Status: Draft
+    Created: 2016-1014
```
Comment (pirapira) [inline EIPS/eip-156.md]: The format of the date is unusual.

File: EIPS/eip-156.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+## Motivation
+The Ethereum network is still to be considered experimental and under heavy development. Bugs that cause incorrectly calculated addresses, funding contracts without code, or simply the lack of replay protection should not lead to end-users losing funds.
+
+## Specification
+The following cases are covered by this proposal and grouped due to the nature of the expected outcome. Note, that each sub-specification needs it's own implementations and tests.
+
+### (156A) Contracts without code
+The below is only usable between the blocks `CONSTANTINOPLE_FORK_BLKNUM` and `CONSTANTINOPLE_FORK_BLKNUM + EIP156_CLAIM_PERIOD` with `EIP156A_CLAIM_PERIOD = 2_000_000` blocks (aproximately one year).
+
+If the `v` value of a signature is (strictly) greater than `1024`, then calculate the sender as follows:
```
Comment (pirapira) [inline EIPS/eip-156.md]: Aren't we using the `v` values for replay protection?

File: EIPS/eip-156.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+2. Let the actual sender be the last 20 bytes of the SHA3 of the lowercase non-prefixed hex-encoded form of `P` instead of the binary raw `P` itself.
+
+### (156B) Incorrectly computed addresses
+Create a Solidity contract with the following functions:
+
+- `declareEmptyContract(index)`: computes `rlp.encode([msg.sender, index])`; if there is a contract at this address with ether and no code, then the contract saves a record that this contract has been checked, and sends an equal amount of _future Ether_ (an ERC20 token) at that account.
+- `declareLowercaseHexAddress(pubkey)`: checks `sha3(pubkey) % 2**160 == msg.sender`; then checks that `sha3(pubkey.encode('hex')) % 2**160` has ether; if it does, then the contract saves a record that this contract has been checked, and sends an equal amount of _future Ether_ (an ERC20 token) at that account
+- `withdraw()`: deletes the `msg.sender`'s _future Ether_, and sends it an equivalent amount of Ether.
```
Comment (pirapira) [inline EIPS/eip-156.md]: Can `declareEmptyContract` (or `declareLowercaseHexAddress`) be called after `withdraw()` again?

File: EIPS/eip-156.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+The hard fork would increase the future Ether contract's balance after `CONSTANTINOPLE_FORK_BLKNUM` by an amount equal to the total quantity of extant _future Ether_.
```
Comment (pirapira) [inline EIPS/eip-156.md]: Can new _future Ether_ be created after the hard fork?

**Comment by tjayrush** [comment]
Please include non-disclosure text with your posts. It's impossible for less-technical people to fully understand implications of comments without explicit disclosures of your interest in this issue.

**Disclosure:** no locked ether in any accounts

**Comment by MicahZoltu** [comment]
A sound argument will be sound regardless of whether the person making it benefits from what they are proposing.  Knowing who has stake and who doesn't is more likely than not to lead to people not listening to sound arguments because they believe the author is "biased".  I would like to keep biases out of this discussion as much as possible (it is already _massively_ biased as it is).

**Comment by MicahZoltu** [comment]
@gcolvin Sorry for not replying sooner, I missed your mention.  The specific rules I think are up for debate and we _should_ have a forum for discussing them.  My primary concern is bloating Ethereum codebase and solving this problem in a complex way when we have an incredibly simple way to solve the same problem (off-chain).  Personally, I would start by just saying something like:

> If you can prove to a reasonable human that you have stuck Ether (that can provably not be retrieved) then your address/amount gets added to the list.  This would account for everything this EIP accounts for, the Parity issue, the issue where the token sale posted the wrong address to their website (bad checksum), ETH sent to account 0, etc.

Others can review submissions and if anyone has a valid objection (e.g., "that isn't actually stuck" or "you can't prove that is stuck") then the address/amount isn't added to the list.

If someone wants, they can write a script that compiles a bunch of these automatically, for others they are "well known" like the Parity multisig bug and such.

**Comment by montagist** [comment]
@tjayrush I think the disclosures make sense/can be helpful, but I disagree that their absence makes understanding "impossible" for non-technical people or anyone else.

**Comment by gzduan** [comment]
Can add a variable in the account, if it is 1 means someone have used the private key of this account , by default, all the variable of new account is 0, if someone used the private key of the account, this variable is automatically into 1. Otherwise the variable will always be 0, if you accidentally sent many tokens to the account that variable is 0 , the system will be in after a period of time, such as 3 years ,if this account still didn't have the records of using the private key, in other word,the variable is still 0. The system can identify the account as a black hole account and the money will be automatically returned.
The mistake of sending money to the black hole address, which is unavoidable, will continue to happen now, and this is the defect of the Ethereum network,
This similar to activate an account,if there have the record of using private key on the blockchain, the account will be automatically activated and will not returned the eth after received the eth three years , the use of paper wallet does is restricted, because there must be activated account operation, but this only to new accounts which private key never used by people , maybe paper wallets are not a necessary a way of storing tokens, Cold wallet signature can be a good choice, can use the private key offline and ensure absolute safety.
Put the tokens sent to the a black hole accounts always happen to us, on the one hand block chain technology ensures the property safety of people, but the careless operation is so despair to ethereum users .the proposal can effectively recover the eth that sent to a black hole account , at the same time ,it also punished the careless users, Because they have to wait at least three years to get the money they lost because of their mistakes.
It's like you have to bind a phone number when you use Facebook account.

**Comment by 5chdn** [comment]
Closing due to the heterogeneous nature of this proposal (156 A, B, C, and D) in favor of EIP #867. Each sub-proposal of the current EIP #156 version could be its own ERP if EIP #867 is accepted.

If EIP #867 is rejected, I don't see much potential for EIP #156 either.

**Pull request closed without merging**